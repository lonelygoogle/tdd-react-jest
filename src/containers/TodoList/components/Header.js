import React, { Component }from 'react';

class Header extends Component {
    constructor (props) {
        super(props)
        this.handleInputChange = this.handleInputChange.bind(this)
        this.handleInputKeyup = this.handleInputKeyup.bind(this)
        this.state = {
            value: ''
        }
    }

    handleInputChange(e) {
        this.setState({
            value: e.target.value
        })
    }

    handleInputKeyup (e) {
        const { value } = this.state
        if (e.keyCode === 13 && value) {
            this.props.addUndoItem(value)
            this.setState({value: ''})
        }
    }

    render () {
        const { value } = this.state
        return (
        <div className="header">
            <div className="header-contain">
                TodoList
                <input
                placeholder="Todo hi"
                className="header-input"
                data-test='input'
                value={value}
                onChange={this.handleInputChange}
                onKeyUp={this.handleInputKeyup}
                />
            </div> 
        </div>
        )
    }
}

export default Header;
