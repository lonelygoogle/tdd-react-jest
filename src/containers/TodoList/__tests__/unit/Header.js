import React from 'react';
// import { render } from '@testing-library/react';
import Header from '../../components/Header';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { findTestWrapper } from '../../../../utils/testUtils'

Enzyme.configure({ adapter: new Adapter() });

describe('Header组件', () => {
    test('渲染样式正常', () => {
      const wrapper = shallow(<Header />)
      expect(wrapper).toMatchSnapshot()
    });
    test('包含input框 初始值为空', () => {
      const wrapper = shallow(<Header />)
      const inputElem = findTestWrapper(wrapper, 'input')
      expect(inputElem.prop('value')).toEqual('')
    });
    
    test('包含input框 跟随用户输入改变', () => {
      const wrapper = shallow(<Header />)
      const inputElem = findTestWrapper(wrapper, 'input')
      const userInput = '今天学习jest'
      inputElem.simulate('change', {
        target: {value: userInput}
      })
      expect(wrapper.state('value')).toEqual(userInput)
    });
    
    test('包含input框 按回车时 如果无内容 无操作', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Header addUndoItem={fn}/>)
      // console.log(wrapper.debug())
      const inputElem = findTestWrapper(wrapper, 'input')
      wrapper.setState({value: ''})
      inputElem.simulate('keyUp', {
        keyCode: 13
      })
      expect(fn).not.toHaveBeenCalled()
      // expect(wrapper.state('value')).toEqual(userInput)
    });
    
    test('包含input框 按回车时 如果有内容 函数被调用', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Header addUndoItem={fn}/>)
      const inputElem = findTestWrapper(wrapper, 'input')
      const userInput = '学习React'
      wrapper.setState({value: userInput})
      inputElem.simulate('keyUp', {
        keyCode: 13
      })
      expect(fn).toHaveBeenCalled()
      expect(fn).toHaveBeenLastCalledWith(userInput)
      
      // expect(wrapper.state('value')).toEqual(userInput)
    });
    
    test('包含input框 按回车时 如果有内容 最后input应该清楚掉', () => {
      const fn = jest.fn()
      const wrapper = shallow(<Header addUndoItem={fn}/>)
      const inputElem = findTestWrapper(wrapper, 'input')
      const userInput = '学习React'
      wrapper.setState({value: userInput})
      inputElem.simulate('keyUp', {
        keyCode: 13
      })
      const newInputElem = findTestWrapper(wrapper, 'input')
      expect(newInputElem.prop('value')).toBe('')
    });
});

