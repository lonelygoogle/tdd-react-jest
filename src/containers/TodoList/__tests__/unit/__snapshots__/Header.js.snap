// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header组件 渲染样式正常 1`] = `ShallowWrapper {}`;

exports[`header渲染样式正常 1`] = `ShallowWrapper {}`;
